### DISCUSSION TITLE
NumericalKinematics problem

### DISCUSSION THREAD

**Discussion by zhanglufeng6789** [visose/Robots#292] [Q&A] [answered]
I would like to know why there are errors in calculating the joint angles of the robot when using the iterative solution, especially when my pose is the initial pose. When I apply the iterative algorithm you provided, I encounter errors in the calculations of the 4th and 6th joint angles. This seems to be a common issue, and I wonder if you could help clarify my confusion.

**Comment by visose** [comment]
Hi, can you give some more info? Are you using it as al library, or a fork adding support to a new robot? How are you "applying the iterative algorithm"? I guess this is not the issue, but just to clarify: numerical solvers always have some error, but hopefully the error is smaller than some threshold that is considered tolerable.

**Reply by zhanglufeng6789** [reply]
Hello, I have identified the current issue, which is due to the singularity problem of the robot. At the initial pose of the robot, which is 0, 0, 0, 0, 0, 0, 0, the joints 4 and 6 of the robot will exhibit typical wrist singularities, resulting in solution errors during iteration.

**Reply by zhanglufeng6789** [reply]
Also, may I ask if there are any good ways to avoid this situation?

**Reply by visose** [reply]
Hi, did you see my previous comment?

**Reply by zhanglufeng6789** [reply]
I can see。I am using the ABB IRB120 robot provided internally by you. In the initial posture, the Newton inverse kinematics calculation was incorrect due to the singularity problem of the robot. I just don't know if there are any good ways to avoid this phenomenon

**Reply by visose** [reply]
If you are using the plugin as an end user, when you use the ABB120, it uses an analytical solver (SphericalWristKinematics). 
As an end user, you would not know which solver is being used on which robot, so I presume you are developing something, not using this project as an end user of the Grasshopper plugin? Could you explain what you are doing in some detail?

**Reply by zhanglufeng6789** [reply]
I am using the robots library for my own learning and secondary development work. I don't want to use the inverse kinematics analysis solution of the robot to calculate now, but to test the effectiveness of the iterative solution. When testing abb irb120 as the object, it was found that when the initial 6-week angles were 0, 0, 0, 0, 0, 0, and 0, although the end pose was within the error range, the axis angles would appear as 0, 0, 0, -0.3, 0, and 0.3, respectively. However, the pose was indeed correct and typical wrist singular points appeared, so the rank of the Jacobian matrix was not full rank. Although pseudo inversion is used as a substitute, the results are often inaccurate. So I want to talk to you about any good ways to avoid it. Thank you

**Reply by visose** [reply, marked as answer]
You can check how it's used with robots that do use the numerical solver, instead of ABBs. It won't attempt to solve in the initial pose with a singularity.
